File: practice/contest/hackerrank/java/Permutation.java
public class Permutation {

	private int [] array  = {1,2,3,4,5,6,7};

	public Permutation(int[] array) {
		this.array = array;
	}

	public void setArray(int []array) {
		this. array = array;  
	}

	public void swap(int i, int j) {
		int tmp = array[i];
		array[i] = array[j];
		array[j] = tmp;
	}

	public void permutate(int left, int right) {
		if (left == right) {
			Utils.printArray(array);
		} else {
			for (int i = left; i <= right; i++) {
				swap(left, i);
				permutate(left + 1, right);
				swap(left, i);
			}
		}
	}

	public void printPermutate(){
		permutate(0, array.length - 1);
	}

}

File: practice/cpp/permutation.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

#define REP(i, a, b) for (int i = a; i <= b; i++)

void swap(char* a, char* b) {
	char *tmp = a;
	*a = *b;
	*b = *tmp;
}

void permutation(char* number, int left, int right) {
	if (left == right) {
		cout << number << endl;
		return;
	}

	REP(i, left, right) {
		swap(number[left], number[i]);
		permutation(number, left + 1, right);
		swap(number[left], number[i]);
	}
}
int main() {
	char str[] = "ABCDEF";
	permutation(str, 0, sizeof(str)/sizeof(char) - 2);
	return 0;
}
